R/elasticsearch_parsers.R

```diff
@@ -390,45 +390,64 @@ unpack_nested_data <- function(chomped_df, col_to_unpack) {
         log_fatal(msg)
     }
     
-    # Avoid side effects
     outDT <- data.table::copy(chomped_df)
-    
-    # Get the column to unpack
     listDT <- outDT[[col_to_unpack]]
     
-    # Make each row a data.table
-    listDT <- lapply(listDT, data.table::as.data.table)
-    
-    # Remove the empty ones... important, due to data.table 1.10.4 bug
-    oldIDs <- which(sapply(listDT, nrow) != 0)
-    listDT <- listDT[oldIDs]
-    
-    # Bind them together with an ID to match to the other data
-    newDT <- data.table::rbindlist(listDT, fill = TRUE, idcol = TRUE)
-    
-    # If we tried to unpack an empty column, fail
-    if (nrow(newDT) == 0) {
+    # Check for empty column
+    if (all(lengths(listDT) == 0)) {
         msg <- "The column given to unpack_nested_data had no data in it."
         log_fatal(msg)
     }
     
-    # Fix the ID because we may have removed some empty elements due to that bug
-    newDT[, .id := oldIDs[.id]]
+    listDT[lengths(listDT) == 0] <- NA
     
-    # Merge
-    outDT[, .id := .I]
-    outDT <- newDT[outDT, on = ".id"]
+    is_df <- purrr::map_lgl(listDT, is.data.frame)
+    is_atomic <- purrr::map_lgl(listDT, is.atomic)
+    is_na <- is.na(listDT)
+    
+    # Bind packed column into one data.table
+    if (all(is_atomic)) {
+        newDT <- data.table::as.data.table(unlist(listDT))
+    } else if (all(is_df | is_atomic)) {
+        # If the packed column contains a mixture of data tables, we need to 
+        # to convert the atomic vectors to data.tables
+        
+        # Find column name to use for NA vectors
+        first_df <- min(which(is_df))
+        col_name <- names(listDT[[first_df]])[1]
+        
+        # Convert non data.frame rows to data.table and assign name to rows
+        # with no name
+        prep_row <- function(x) {
+            if (is.atomic(x)) {
+                x <- data.table::as.data.table(x)
+                if (is.na(x)) names(x) <- col_name
+                else names(x) <- col_to_unpack
+            } 
+            x
+        }
+        newDT <- purrr::map(listDT, prep_row)
+        
+        newDT <- data.table::rbindlist(newDT, fill = TRUE)
+    } else {
+        msg <- paste0("Each row in column ", col_to_unpack, " must be a data frame or a vector.")
+        futile.logger::flog.fatal(msg)
+        stop(msg)
+    }
     
-    # Remove the id column and the original column
-    outDT <- outDT[, !c(".id", col_to_unpack), with = FALSE]
+    # Create the unpacked data.table by replicating the originally unpacked
+    # columns by the number of rows in each entry in the original unpacked column
+    times_to_replicate <- pmax(purrr::map_int(listDT, NROW), 1)
+    # Replicate the rows of the data.table by entries of times_to_replicate but drop col_to_unpack
+    replicatedDT <- chomped_df[rep(1:nrow(chomped_df), times_to_replicate)]
```

Comment:
Makes sense, though perhaps you should test out my former strategy of using the built-in idcol argument of rbindlist. That's essentially replicating the logic rep(1:nrow(chomped_df), times_to_replicate) but I suspect it will be faster since it's data.table haha

Reply:
So I tried benchmarking a few versions of this function with a large `listDT` that was full of data.tables (no atomic):

1) the old function
2) the old function with `lapply(listDT, data.table::as.data.table)` commented out
3) this version

(3) was twice as fast as (1), but (2) was twice as fast as (3). I suspect the difference is all the copies you're making in this chunk.

Reply:
Yep, I think these changes lost the speedup. The main speedup was from dropping `data.table::as.data.table`.

Reply:
Try out this update's speed. For what it's worth I tried out this function where I used the same merge as the original function instead of `rep` and it ended up being about 20% slower. I think that makes sense because I'm still using a data.table `[`. So it shouldn't copy the data any more than a data.table merge would copy the data. It's basically going through the same algorithm as merging in the `data.table` with `.id` being the same as `times_to_replicate`. Furthermore, I think it makes sense this way since we don't have to deal with creating a `.id` column which could cause a problem if `chomped_df` has a `.id` column.

Reply:
Hey, thanks for your patience here. No idea why your build is failing. It may just be transient.

Just tested this update's speed on my test dataset again (which has 1000 data frames in the `col_to_unpack`, no atomic or lists or NAs). The numbers below are the median times each function took on my computer.
- Old function: 153 ms
- Old function with the `lapply(listDT, data.table::as.data.table)` commented out: 69 ms
- Your first version: 97 ms
- Your second version: 89 ms
- A hybrid: 17 ms

For the hybrid, I built off your second version. I changed the top to:
```
inDT <- data.table::copy(chomped_df)
listDT <- inDT[[col_to_unpack]]
inDT[, (col_to_unpack) := NULL]
inDT[, .id := .I]
```
then added the `idcol = TRUE` back to `rbindlist` and merged after `rbindlist` like so:
```
outDT <- inDT[newDT, on = ".id"]
outDT[, .id := NULL]
return(outDT)
```

So I still think there's value in researching `idcol` as an option. If you're worried about the `.id` column restriction, we could have that join column be a random UUID or something for safety. You can, for example, use `idcol = myUUIDvar` instead of `idcol = TRUE` to automatically name that ID column.